projs/shadow/shadow-engine/core/src/profile/Profiler.cpp

```diff
@@ -0,0 +1,556 @@
+#include <shadow/profile/Profiler.h>
+#include <vector>
+#include "shadow/assets/fs/iostream.h"
+#include "shadow/util/Syncronization.h"
+#include "shadow/core/Time.h"
+#include <shadow/core/Thread.h>
+#include <shadow/platform/Common.h>
+
+#ifdef _WIN32
+#define INITGUID
+#define NOGDI
+#define WIN32_LEAN_AND_MEAN
+#include <Windows.h>
+#include <evntcons.h>
+#include <atomic>
+#include <cstring>
+
+#endif
+
+namespace SH {
+
+  // Allows the Profiler to be used on multiple threads simultaneously, separating the data collected.
+  // Every thread has its own set of open blocks, its own mutex, and its own visibility toggles.
+  // Consider the ThreadContext as a horizontal bar on a profiler menu, showing what one thread is doing at any given time.
+  struct ThreadContext {
+
+    ThreadContext(size_t bufferSize) : oms(), openBlocks() {
+        oms.resize(bufferSize);
+        openBlocks.reserve(64);
+    }
+
+    std::vector<uint32_t> openBlocks;
+    // Used as the data buffer.
+    ShadowEngine::OutputMemoryStream oms;
+    size_t begin = 0;
+    size_t end = 0;
+    ShadowEngine::Mutex mut;
```

Comment:
This is also missing

Reply:
https://github.com/Candle-Fire/umbra/pull/31/files#diff-ca5ea1c9ce87bda00ae365c87da348d6a1ce7b7a37820d481913659365389c0fR11